packages/driver/src/cy/commands/querying.js

```diff
@@ -0,0 +1,634 @@
+        // and we have been explictly told to filter
+        // then just attempt to filter out elements from our within subject
+        if (!$el.length && options.withinSubject && options.filter) {
+          const filtered = options.withinSubject.filter(selector)
+
+          // reset $el if this found anything
+          if (filtered.length) {
+            $el = filtered
+          }
+        }
+
+        // store the $el now in case we fail
+        setEl($el)
+
+        // allow retry to be a function which we ensure
+        // returns truthy before returning its
+        if (_.isFunction(options.onRetry)) {
+          const ret = options.onRetry.call(ctx, $el)
+
+          if (ret) {
+            log($el)
+
+            return ret
+          }
+        } else {
+          log($el)
+
+          return $el
+        }
+      }
+
+      resolveElements = () => {
+        return Promise.try(getElements).then(($el) => {
+          if (options.verify === false) {
+            return $el
+          }
+
+          return cy.verifyUpcomingAssertions($el, options, {
+            onRetry: resolveElements,
+          })
+        })
+      }
+
+      return resolveElements()
+    },
+
+    root (options = {}) {
+      _.defaults(options, { log: true })
+
+      if (options.log !== false) {
+        options._log = Cypress.log({ message: '' })
+      }
+
+      const log = function ($el) {
+        if (options.log) {
+          options._log.set({ $el })
+        }
+
+        return $el
+      }
+
+      const withinSubject = cy.state('withinSubject')
+
+      if (withinSubject) {
+        return log(withinSubject)
+      }
+
+      return cy.now('get', 'html', { log: false }).then(log)
+    },
+  })
+
+  Commands.addAll({ prevSubject: ['optional', 'window', 'document', 'element'] }, {
+    contains (subject, filter, text, options = {}) {
+      // nuke our subject if its present but not an element.
+      // in these cases its either window or document but
+      // we dont care.
+      // we'll null out the subject so it will show up as a parent
+      // command since its behavior is identical to using it
+      // as a parent command: cy.contains()
+      if (subject && !$dom.isElement(subject)) {
+        subject = null
+      }
+
+      if (_.isRegExp(text)) { // .contains(filter, text)
+        // Do nothing
+      } else if (_.isObject(text)) { // .contains(text, options)
+        options = text
+        text = filter
+        filter = ''
+      } else if (_.isUndefined(text)) { // .contains(text)
+        text = filter
+        filter = ''
+      }
+
+      _.defaults(options, { log: true })
+
+      if (!(_.isString(text) || _.isFinite(text) || _.isRegExp(text))) {
+        $utils.throwErrByPath('contains.invalid_argument')
+      }
+
+      if (_.isBlank(text)) {
+        $utils.throwErrByPath('contains.empty_string')
+      }
+
+      const getPhrase = function (type, negated) {
```

Comment:
These arguments are not used

```suggestion
      const getPhrase = function () {
```